x-pack/plugins/task_manager/task_store.ts

```diff
@@ -96,54 +103,86 @@ export class TaskStore {
-          status: { type: 'keyword' },
-          params: { type: 'text' },
-          state: { type: 'text' },
-          user: { type: 'keyword' },
-          scope: { type: 'keyword' },
-        },
-      },
-    };
+    let existingVersion = -Infinity;
+    const templateName = this.index;
+    const templateVersion = getTemplateVersion(xPackage.version);
 
     try {
-      const templateResult = await this.callCluster('indices.putTemplate', {
-        name: this.index,
-        body: {
-          index_patterns: [this.index],
-          mappings: {
-            _doc: {
-              dynamic: 'strict',
-              properties,
-            },
-          },
-          settings: {
-            number_of_shards: 1,
-            auto_expand_replicas: '0-1',
-          },
-        },
+      // check if template exists
+      const templateCheck = await this.callCluster('indices.getTemplate', {
+        name: templateName,
+        filter_path: '*.version',
       });
-      this._isInitialized = true;
-      return templateResult;
+      // extract the existing version
+      existingVersion = templateCheck[templateName] ? templateCheck[templateName].version : null;
     } catch (err) {
-      throw err;
+      if (err.statusCode !== 404) {
+        throw err; // ignore not found
+      }
+    }
+
+    if (existingVersion > templateVersion) {
+      throw new Error(
+        `Template version ${templateVersion} could not be saved: existing template version ${existingVersion} is newer!`
+      );
+    } else if (existingVersion === templateVersion) {
+      this.logger.debug(
+        `Not installing ${this.index} index template: version ${templateVersion} already exists.`
+      );
+      return; // the latest template is already saved. Nothing to do.
+    } else if (existingVersion < templateVersion) {
+      this.logger.info(
+        `Upgrading ${
+          this.index
+        } index template. Old version: ${existingVersion}, New version: ${templateVersion}.`
+      );
     }
 
-    return;
+    const templateResult = await this.callCluster('indices.putTemplate', {
+      name: templateName,
+      body: {
+        index_patterns: [this.index],
+        mappings: {
+          _doc: {
+            dynamic: 'strict',
+            properties: {
+              type: { type: 'keyword' },
+              task: {
+                properties: {
+                  taskType: { type: 'keyword' },
+                  runAt: { type: 'date' },
+                  interval: { type: 'text' },
+                  attempts: { type: 'integer' },
+                  status: { type: 'keyword' },
+                  params: { type: 'text' },
+                  state: { type: 'text' },
+                  user: { type: 'keyword' },
+                  scope: { type: 'keyword' },
+                },
+              },
+            },
+          },
+        },
+        settings: {
+          number_of_shards: 1,
+          auto_expand_replicas: '0-1',
```

Comment:
Seems like there wouldn't be downsides, right? It's fine with me.

Reply:
It reduces the on-disk impact, but it technically increases the CPU cost (by a trivial amount) to use the better compression.

Reply:
I am fine with not doing it, because if users need it they can just set it on their own cluster, right?

I hope I'm not cursing myself by saying this, but I think the `.kibana_task_manager` index will stay pretty small. It doesn't keep historical data from completed tasks, so the growth will come from scheduling many tasks, and recurring tasks that grow their state (text data) over time.

Reply:
We will eventually be keeping historical runs but probably prudent to use another index for that if historical tracking is enabled.

I would expect the index to grow with usage and time since you could have a user do something like nagios style alerts eventually. Which might become significant if they are monitoring uptime of a significant number of machines for instance. So i'd just assume the index is going to get large. Unless we are leaning away from using task_manager for this type of purpose in which case I need to just start writing my own implementation of background processing for alerts.

Reply:
++ for a separate, time-based index to store result history. That was an intentional design choice for Task Manager: only store what it needs to care about. Client code has access to the return data from the task worker function, and can preserve that into a time-based index. Maybe you'd want that to be done in bulk though? Or go through a Beat? It could be a lot of data.